lib/rules/no-unused-properties.js

```diff
@@ -0,0 +1,167 @@
+/**
+ * @fileoverview Disallow unused properties, data and computed properties.
+ * @author Learning Equality
+ */
+'use strict'
+
+// ------------------------------------------------------------------------------
+// Requirements
+// ------------------------------------------------------------------------------
+
+const remove = require('lodash/remove')
+const utils = require('../utils')
+
+// ------------------------------------------------------------------------------
+// Constants
+// ------------------------------------------------------------------------------
+
+const GROUP_PROPERTY = 'props'
+const GROUP_DATA = 'data'
+const GROUP_COMPUTED_PROPERTY = 'computed'
+const GROUP_WATCHER = 'watch'
+
+const PROPERTY_LABEL = {
+  [GROUP_PROPERTY]: 'property',
+  [GROUP_DATA]: 'data',
+  [GROUP_COMPUTED_PROPERTY]: 'computed property'
+}
+
+// ------------------------------------------------------------------------------
+// Helpers
+// ------------------------------------------------------------------------------
+
+/**
+ * Extract names from references objects.
+ */
+const getReferencesNames = references => {
+  if (!references || !references.length) {
+    return []
+  }
+
+  return references.map(reference => {
+    if (!reference.id || !reference.id.name) {
+      return
+    }
+
+    return reference.id.name
+  })
+}
+
+/**
+ * Report all unused properties.
+ */
+const reportUnusedProperties = (context, properties) => {
+  if (!properties || !properties.length) {
+    return
+  }
+
+  properties.forEach(property => {
+    context.report({
+      node: property.node,
+      message: `Unused ${PROPERTY_LABEL[property.groupName]} found: "${property.name}"`
+    })
+  })
+}
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'disallow unused properties, data and computed properties',
+      category: undefined,
+      url: 'https://eslint.vuejs.org/rules/no-unused-properties.html'
+    },
+    fixable: null,
+    schema: []
+  },
+
+  create (context) {
+    let hasTemplate
+    let rootTemplateEnd
+    let unusedProperties = []
+    const thisExpressionsVariablesNames = []
+
+    const initialize = {
+      Program (node) {
+        if (context.parserServices.getTemplateBodyTokenStore == null) {
+          context.report({
+            loc: { line: 1, column: 0 },
+            message:
+              'Use the latest vue-eslint-parser. See also https://vuejs.github.io/eslint-plugin-vue/user-guide/#what-is-the-use-the-latest-vue-eslint-parser-error.'
+          })
+          return
+        }
+
+        hasTemplate = Boolean(node.templateBody)
+      }
+    }
+
+    const scriptVisitor = Object.assign(
+      {},
+      {
+        'MemberExpression[object.type="ThisExpression"][property.type="Identifier"][property.name]' (
```

Comment:
Could you change to support the destructuring assignment?

```js
const {prop} = this
```

Also could you change to support the track the `this`?
I want to support assignment expressions as follows:

```js
const vm = this
vm.prop
```